### DISCUSSION TITLE
Run Mbi6033 LED/Chip IC with WLED

### DISCUSSION THREAD

**Discussion by seyfettinkazanir** [Makuna/NeoPixelBus#715] [Q&A]
I want to use neopixelbus with mbi6033 ic. I know there is a library that wled uses. They said to a question I asked from there, you can do it with neopixelbus, you have to ask to edit it.

I want to run mbi6033 with wled but I don't know how to proceed. I need help.
Mbi6033 is actually an economical solution. It runs 4 channels with 1 chip. It is an integrated that uses spi. I would like to share what we know about the mbi6033.

It has 2 different modes. with current Gain and without current gain. I want to use it without. Spi data contains a 48-bit header, and its channels consist of 16-bit and gray scale data. We have the header information. There is also a data string of a working device taken with a logic analyzer. If you need more information, please don't hesitate to ask.

[MBI6033 Application Note V1.00-EN.PDF](https://github.com/Makuna/NeoPixelBus/files/11724986/MBI6033.Application.Note.V1.00-EN.PDF)

![pic2](https://github.com/Makuna/NeoPixelBus/assets/27890571/8440f2af-cb74-4312-9c50-b559f2c1ac37)
![pic1](https://github.com/Makuna/NeoPixelBus/assets/27890571/fdac01a7-feac-41af-b4a2-bab43f591f1b)

**Comment by Makuna** [comment]
What uC are you using?

The issue here is that it uses the CLK to do the reset.  Specifically, it wants to send a single 0 bit with delay before and after.  This is not achievable with hardware SPI as it sends data by bytes and can't send a single bit.  Can I manually toggle it when the hardware SPI controls the PIN, that can only be answered by trying.

With BitBang (which I support and implemented) theatrically it could be modified to send a single bit.

There is also this timing stuff mentioned based on number chips, but DotStars (3 wire chips like this that have Clk, Data, Gnd) this isn't uncommon, but it requires special coding for it.

The rest (header and 16bit elements) is already supported it would just need its header to be exposed as a feature setting (other chips have this also).

**Reply by seyfettinkazanir** [reply]
I use uC esp32 and some times wt01-eth (lan + esp32).

**Comment by seyfettinkazanir** [comment]
I read a topic about software spi in two-line chips, would it help?  topic https://github.com/Aircoookie/WLED/issues/2542 related setting: https://github.com/Aircoookie/WLED/issues/2542#issuecomment-1465017395

**Reply by Makuna** [reply]
Not really.  The issue is the errant pulse to the clock with the timing before and after to force the chips to "reset".  This process of reset (using a long delay at the end) is normal, but the required clock pulse is not. And its not a feature of SPI so it has to be tacked on.

**Reply by seyfettinkazanir** [reply]
So how to do it, can you help me?  I'm not sure I know how to edit. 

Should I edit the spi settings in the neopixelbus library?

**Reply by Makuna** [reply]
If you want to check out the rabbit hole:
Look at Sm1617 support (NeoSm168xxFeature.h & Sm1676GenericMethod.h) as an example of building custom settings and DotStarGenericMethod.h for the extra data based on count.

If you can wait, I will take a look at a frame work method for it just to test out that extra weird clock bit used for reset.

**Reply by seyfettinkazanir** [reply]
Thanks for your help, I'm waiting.

**Reply by Makuna** [reply]
Do you what LEDs are attached and what order?

Since each chip has 8 channels, this could be 3 RGB, or this could be 2 RGBW.

**Reply by seyfettinkazanir** [reply]
[mbı6033.pdf](https://github.com/Makuna/NeoPixelBus/files/11754824/mbi6033.pdf)

our design led strip is like this.

**Reply by Makuna** [reply]
From your circuit design, the color order is GRB, so use `NeoGrb48Feature`.
The first pixel will be group 3, so C3,B3,A3 from your circuit.

**Comment by Makuna** [comment]
https://github.com/Makuna/NeoPixelBus/tree/Mbi6033
This branch exposes a new method set for Mbi6033.  You can clone this directly into your arduino libraries (see readme).

Assuming RGB pixel order in the following examples.  (updated: 16 bits per element uses NeoRgb48Feature)

For bit bang, use Mbi6033Method like
```
NeoPixelBus<NeoRgb48Feature, Mbi6033Method> strip(PixelCount, ClockPin, DataPin);
```
For hardware SPI, use Mbi6033SpiMethod method like 
```
NeoPixelBus<NeoRgb48Feature, Mbi6033SpiMethod> strip(PixelCount);
```
and if your platform is ESP32 the pins for the hardware SPI are define when you call Begin() like
```
strip.Begin(sck, miso, mosi, ss);
```

**Reply by seyfettinkazanir** [reply]
NeoPixelBus<NeoRgb48Feature, Mbi6033SpiMethod> strip(PixelCount);

I don't know if I did it right, but I got a data using this section. I am sending an attachment. I am also attaching the sample code.

![pic11](https://github.com/Makuna/NeoPixelBus/assets/27890571/12e3c91b-e1b9-4649-a7d5-69f103f02438) logic anlyzer pic.

[Yeni Metin Belgesi.txt](https://github.com/Makuna/NeoPixelBus/files/11756894/Yeni.Metin.Belgesi.txt)
![WhatsApp Image 2023-06-15 at 14 24 34](https://github.com/Makuna/NeoPixelBus/assets/27890571/aca9e4de-4ed5-4aab-b711-a568a956e7b4)  this is one mbi6033 strip

**Reply by seyfettinkazanir** [reply]
I think there is a deviation in the clock signal. approximately twice as long.
![11111](https://github.com/Makuna/NeoPixelBus/assets/27890571/c7a5feb1-a847-41d3-a49d-43973ef76910)

**Reply by Makuna** [reply]
Can you capture what you see before the header?  Trying to see the reset clock pulse.

Also, let's try to slow it down some, so use `Mbi6033Spi2MhzMethod`, by default is trying to send at 10Mhz.

**Comment by Makuna** [comment]
The clock width shouldn't be a problem, this is why they have a clock signal as it is used to time and space bits.  It can vary.  If you use hardware SPI, it usually is pretty solid though.

You can't reuse the same SPI for two different busses like your sample is showing.  The Arduino SPI object is used, it is a singleton and a single SPI must be dedicated to the NeoPixelBus.  This statement from your sketch enables both at the same time.
```
#define USE_DEFAULT_SPI_PORT 1
#define USE_ALTERNATE_SPI_PORT 1
```
Further, the capture states 24 chips are being configured, but your sketch states 4 pixels or just One Chip.  Are you sure you have the right sketch and capture?

I want to impress upon you to keep your initial test sketch small and clean as possible to help me read it and help us diagnose the issue.

**Comment by Makuna** [comment]
Ok, I tested that hardware SPI will not allow the manipulation of the clock to provide the needed reset, so this means only bitbang is supported.  I pushed some fixes to the above branch so pick up those changes.

So, declare and construct your bus like...
```
NeoPixelBus<NeoGrb48Feature, Mbi6033Method> strip(PixelCount, ClockPin, PixelPin);
```
And here is a capture showing the reset at the front for a single pixel (but still outputs the full chip of 4 pixels).
![image](https://github.com/Makuna/NeoPixelBus/assets/7972357/819163a8-91c0-40f9-b960-cebfbc874b40)

**Reply by seyfettinkazanir** [reply]
I tried to create a file using the files you posted but the compilation failed. I am attaching the error message.

In file included from .pio/libdeps/esp32dev/NeoPixelBus/src/internal/NeoMethods.h:38,
                 from .pio/libdeps/esp32dev/NeoPixelBus/src/NeoPixelBus.h:43,
                 from src/main.cpp:10:
.pio/libdeps/esp32dev/NeoPixelBus/src/internal/methods/Mbi6033GenericMethod.h: In instantiation of 'void Mbi6033MethodBase<T_TWOWIRE>::Initialize(int8_t, int8_t, int8_t, int8_t) [with T_TWOWIRE = TwoWireBitBangImple; int8_t = signed char]':
.pio/libdeps/esp32dev/NeoPixelBus/src/NeoPixelBus.h:106:9:   required from 'void NeoPixelBus<T_COLOR_FEATURE, T_METHOD>::Begin(int8_t, int8_t, int8_t, int8_t) [with T_COLOR_FEATURE = NeoGrb48Feature; T_METHOD = Mbi6033MethodBase<TwoWireBitBangImple>; int8_t = signed char]'
src/main.cpp:93:67:   required from here
.pio/libdeps/esp32dev/NeoPixelBus/src/internal/methods/Mbi6033GenericMethod.h:73:9: error: no matching 
function for call to 'TwoWireBitBangImple::begin(int8_t&, int8_t&, int8_t&, int8_t&)'
         _wire.begin(sck, miso, mosi, ss);
         ^~~~~
In file included from .pio/libdeps/esp32dev/NeoPixelBus/src/internal/methods/DotStarGenericMethod.h:33,                 from .pio/libdeps/esp32dev/NeoPixelBus/src/internal/NeoMethods.h:31,
                 from .pio/libdeps/esp32dev/NeoPixelBus/src/NeoPixelBus.h:43,
                 from src/main.cpp:10:
.pio/libdeps/esp32dev/NeoPixelBus/src/internal/methods/TwoWireBitBangImple.h:49:10: note: candidate: 'void TwoWireBitBangImple::begin()'
     void begin()
          ^~~~~
.pio/libdeps/esp32dev/NeoPixelBus/src/internal/methods/TwoWireBitBangImple.h:49:10: note:   candidate expects 0 arguments, 4 provided

**Reply by seyfettinkazanir** [reply]
no matching function for call to 'TwoWireBitBangImple::begin(int8_t&, int8_t&, int8_t&, int8_t&)' this is show error message.

**Reply by Makuna** [reply]
Calling Begin(clock, miso, mosi, cs) is only available and required for ESP32 Hardware SPI.  

Since we can't use hardware SPI as it won't let us send a single bit for the reset, then the clock and mosi (data out) are supplied at the constructor (as the example above has) and the you need to remove the arguments to begin.

**Comment by seyfettinkazanir** [comment]
I was able to compile the code with the edits you specified. We have now observed that the data works. What kind of a way should be followed to add mbi6033 to wled with this prepared method. We observed that the first chip works when pixelcount is 4, and the second chip works when it is 8. I am sending a working video attached. Thank you for your help. 

https://github.com/Makuna/NeoPixelBus/assets/27890571/e2ed480a-4990-4595-acd5-6f54fbbf6a9d


https://github.com/Makuna/NeoPixelBus/assets/27890571/9f30b34e-76b1-4057-96f1-6ce0c34e7a26

**Reply by seyfettinkazanir** [reply]
I'm trying to figure out how it works with a few changes.

pixelcount = 8 ;

     strip.SetPixelColor(0, red);
     strip.SetPixelColor(1, green);
     strip.SetPixelColor(2, blue);
     strip.SetPixelColor(3, white);
     strip.SetPixelColor(4, red);
     strip.SetPixelColor(5, green);
     strip.SetPixelColor(6, blue);
     strip.SetPixelColor(7, white);

Working video is attached.


https://github.com/Makuna/NeoPixelBus/assets/27890571/1f91a2d4-58af-4e96-9c81-824461cf1758

**Reply by seyfettinkazanir** [reply]
I managed to use it by integrating it into the programs in the sample files. animation images are very nice. I am attaching an animation. Same chip but a different strip.

https://github.com/Makuna/NeoPixelBus/assets/27890571/7e06ab1c-5945-4b47-ad2e-523d3c09c85c

**Comment by Makuna** [comment]
> I'm trying to figure out how it works with a few changes.
> 
> pixelcount = 8 ;
> 
> ```
>  strip.SetPixelColor(0, red);
>  strip.SetPixelColor(1, green);
>  strip.SetPixelColor(2, blue);
>  strip.SetPixelColor(3, white);
>  strip.SetPixelColor(4, red);
>  strip.SetPixelColor(5, green);
>  strip.SetPixelColor(6, blue);
>  strip.SetPixelColor(7, white);
> ```

Of your videos, which end is the first pixel?  From your code snippit, it would be the far end but it looks like those aren't lit.

I just want make sure you have the right color order so if you set the first pixel to red, you get red.

If everything looks good, then I will review the code for any cleanup and merge it into master today.

**Reply by seyfettinkazanir** [reply]
![IMG-20230617-WA0022](https://github.com/Makuna/NeoPixelBus/assets/27890571/ac4da513-9ca4-4e83-a6ce-074227f2b6e7)

I think the first data is the last data of the spi data.  I guess it's the opposite.

Where we log in with the white led esp32.  the direction of the line is from white to red.  In fact, I took the video from the end of the line in the 4 led video I sent last.

**Reply by Makuna** [reply]
Oh this means something is still wrong.  The key problem is that White (no matter what color order for RGB pixels) should not change location unless it's really a RGBW pixel.  
Could you try this for me?
```
 strip.SetPixelColor(0, red);
 strip.SetPixelColor(1, green);
 strip.SetPixelColor(2, blue);
 strip.SetPixelColor(3, black);
 strip.SetPixelColor(4, red);
 strip.SetPixelColor(5, green);
 strip.SetPixelColor(6, blue);
 strip.SetPixelColor(7, white);
```
And post a picture.

**Reply by seyfettinkazanir** [reply]
I am sending a video with the code you requested. Where the first spi data is sent in the section where the white led is lit. So uC is sending data from here.


https://github.com/Makuna/NeoPixelBus/assets/27890571/1bdff220-0404-4b74-a815-60412bb04bb3

**Reply by Makuna** [reply]
Interesting. Not only is the order of pixels on the chip taken in reverse, but the order of the data for chips is read in reverse.  So the first physical chip uses the last sent data.  Since both are reversed, as long as you match the strip size correctly, it works fine if you consider the opposite end from the connection as the first pixel.

Is this a problem for you?

**Reply by seyfettinkazanir** [reply]
I think it will not be a problem since we will use it with wled. Because there is a reverse operation feature. We can use it.

**Comment by Makuna** [comment]
BTW:  These chips support 16 bits per color element, so you can use Rgb48Color object to define your colors with more resolution.

**Reply by seyfettinkazanir** [reply]
I remember that I used rgb48color in my last animated video. I can check and let you know.

**Reply by Makuna** [reply]
Like...
```
const Rgb48Color white(0xffff, 0xffff, 0xffff);
const Rgb48Color red(0xffff, 0x0000, 0x0000);
etc
```

**Reply by seyfettinkazanir** [reply]
I changed it like this on your suggestion.  created using these settings.  The video was shot with these settings.

`const Rgb48Color red(0xffff,0x0000,0x0000);`
`const Rgb48Color green(0x0000,0xffff,0x0000);`
`const Rgb48Color blue(0x0000,0x0000,0xffff);`
`const Rgb48Color white(0xffff,0xffff,0xffff);`
`const Rgb48Color black(0);`

https://github.com/Makuna/NeoPixelBus/assets/27890571/64431835-0039-44a1-ab76-153ba8b98a33

**this is orijinal.**

`#define colorSaturation 128`
`RgbColor red(colorSaturation, 0, 0);`
`RgbColor green(0, colorSaturation, 0);`
`RgbColor blue(0, 0, colorSaturation);`
`RgbColor white(colorSaturation);`
`RgbColor black(0);`

This is how I was using it before your suggestion. I guess both are the same thing. I think the saturation value adjusts the light level.

**I edited it like this.** 

` #define colorSaturation 8000` 
`Rgb48Color red(colorSaturation, 0, 0);` 
`Rgb48Color green(0, colorSaturation, 0);` 
`Rgb48Color blue(0, 0, colorSaturation);` 
`Rgb48Color white(colorSaturation);` 
`Rgb48Color black(0);`

**Reply by seyfettinkazanir** [reply]
I sent data with a different strip so that it is easy to observe. mbi6033 is still the same.

![WhatsApp Image 2023-06-19 at 11 07 10](https://github.com/Makuna/NeoPixelBus/assets/27890571/c19dd634-4c85-42f5-b94c-97fcc9b8f3b8)

**Reply by Makuna** [reply]
If you don't care about the higher resolution of 16 bits per element, you can still use the RgbColor object.  Internally it will auto convert to a Rgb48Color when you call SetPixelColor. Note that calling GetPixelColor will return a Rgb48Color, which will not auto convert to RgbColor (due to loss of information this isn't automatic).  You can just call RgbColor( rgb48value) to convert down.

**Reply by seyfettinkazanir** [reply]
RgbColor seems to be working for now, but it would be nice to apply the changes anyway. Which part of the code do I add?

**Reply by Makuna** [reply]
Nothing to change other than just use a Rgb48Color to set the color.  

I will merge in the changes as they are with the note about the reverse order.

**Reply by seyfettinkazanir** [reply]
So what should I do to add mbi6033 to Wled? will it be added automatically from the library? I wanted to edit the interface with vscode but I haven't been successful yet.

**Reply by Makuna** [reply]
I am not an expert on WLED, but the Key is the Feature and Method names.  This is a good question to ask the WLED discussion, but I believe it just some minor edits to config files.  Remember to be specific that it's a 16bit per element color object, as this support was new this year.

NeoGrb48Feature : look for NeoGrb48Feature/NeoRgb48Feature first to see if its already supported,  otherwise look for NeoGrbFeature as an exmple to copy. 
Mbi6033Method : look for DotStarMethod as an example to copy.

**Reply by seyfettinkazanir** [reply]
such a change? I think it needs to be added to the bus wrapper file. 

`#ifdef WLED_USE_ETHERNET`
`// fix for #2542 (by @BlackBird77)`
`#define B_HS_DOT_3 NeoPixelBusLg<DotStarBgrFeature, DotStarEsp32HspiHzMethod, NeoGammaNullMethod> //hardware HSPI (was DotStarEsp32DmaHspi5MhzMethod in NPB @ 2.6.9)`
`#else`
`#define B_HS_DOT_3 NeoPixelBusLg<DotStarBgrFeature, DotStarSpiHzMethod, NeoGammaNullMethod> //hardware VSPI`
`#endif`
`#define B_SS_DOT_3 NeoPixelBusLg<DotStarBgrFeature, DotStarMethod, NeoGammaNullMethod>    //soft SPI`

**Reply by seyfettinkazanir** [reply]
I haven't got any results yet while working on these. 

`//APA102`
`#ifdef WLED_USE_ETHERNET`
`// fix for #2542 (by @BlackBird77)`
`#define B_HS_DOT_3 NeoPixelBusLg<DotStarBgrFeature, DotStarEsp32HspiHzMethod, NeoGammaNullMethod> //hardware HSPI (was DotStarEsp32DmaHspi5MhzMethod in NPB @ 2.6.9)`
`#else`
`#define B_HS_DOT_3 NeoPixelBusLg<DotStarBgrFeature, DotStarSpiHzMethod, NeoGammaNullMethod> //hardware VSPI`
`#endif`
`#define B_SS_DOT_3 NeoPixelBusLg<NeoRgbFeature, Mbi6033Method, NeoGammaNullMethod>`
`//#define B_SS_DOT_3 NeoPixelBusLg<DotStarBgrFeature, DotStarMethod, NeoGammaNullMethod>    //soft SPI`


Do you have a chance to look at it with a logic analyzer?

**Reply by Makuna** [reply]
Not with WLED.  But under Arduino IDE yes.  I don't build WLED myself.  Again, their discussion forum is the best place to ask about enhancing it with new NeoPixelBus methods and features.

**Reply by seyfettinkazanir** [reply]
How can I understand the main differences between NeoRgbFeature and NeoRgb48Feature. I want to add it to WORD but I guess it doesn't work very well unless it's 16 bit. I am getting an error during compilation. I'm sending a photo attached, I think I'm still stuck somewhere about the compatibility of the part with neopixelbus. 
![1](https://github.com/Makuna/NeoPixelBus/assets/27890571/c5ace601-6099-439c-be0e-9391014dd336)

**Reply by Makuna** [reply]
The key difference beyond 8bit/16bit internals is the ColorObject, NeoRgbFeature uses a RgbColor object.  NeoRgb48Feature uses a Rgb48Color.  

You would have to look at the lines the error is showing on (bus_wrapper.h line 991 & 992) to see what its doing.  I suspect it's trying to assignment like rgbColor = rgb48Color, which you can't do as its a loss of information.  

Check with the WLED guys, do they support 16bit per color elements and how to use it.

Did you search for Rgb48Feature to see if they had an example?

**Reply by seyfettinkazanir** [reply]
removed these two lines of error from the code. I compile with NeoRgb48Feature, but since the 2nd byte is assigned as 0x00, it seems to be dim all the time.

  `//case` I_HS_DOT_3: col = (static_cast<B_HS_DOT_3>(busPtr))->GetPixelColor(pix); break; `
   ` //case I_SS_DOT_3: col = (static_cast<B_SS_DOT_3>(busPtr))->GetPixelColor(pix); break; `

Could it be related to Fastled? I guess data preparations such as color selection were done there.
I wrote in forums with wled powers. I'm waiting for someone's reply. Someone referred to the official, I'm waiting for a reply. nickname @blazoncek.


![16bit4 white data](https://github.com/Makuna/NeoPixelBus/assets/27890571/25082cf6-06c0-4d58-818b-fa86f9898b1e)
![16bit3](https://github.com/Makuna/NeoPixelBus/assets/27890571/91e0f0ce-c0da-4261-9400-23f528864132)
![16bit2](https://github.com/Makuna/NeoPixelBus/assets/27890571/10fb62ab-993d-4934-8874-8938d60ebb69)
![16bit1](https://github.com/Makuna/NeoPixelBus/assets/27890571/eb9de59b-2126-4b4e-ada6-ff5cf7854da1)

**Reply by seyfettinkazanir** [reply]
Such a statement has been said on the forum. "WLED operates on 8-bit channels but it supports UCS890x 16-bit chipsets." Is it right to go this way to reach the solution? 

[(https://www.led-stuebchen.de/download/UCS8903_English%202017-8-22.pdf)]

![data1](https://github.com/Makuna/NeoPixelBus/assets/27890571/3e37f8ec-cf86-4de0-a385-2359ae75bde4)

The data structure is 48 but I don't know exactly.

**Reply by blazoncek** [reply]
WLED uses on the fly conversion from internal 8bit color representation to 16bit.
As this chipset has been merged to NPB only 2 days ago, it will take some time to include it in WLED, if at all.
@seyfettinkazanir open an issue on WLED GH requesting support for the chipset and include released version and appropriate method of NPB where support has been added.

**Reply by seyfettinkazanir** [reply]
@blazoncek Thank you very much for your quick turnaround. I wrote the comment when "softhack" told me to reach you. I started a topic on the subject.  https://wled.discourse.group/t/run-mbi6033-led-chip-ic-with-wled/8569

**Comment by Makuna** [comment]
https://github.com/Makuna/NeoPixelBus/issues/718 tracking issue
https://github.com/Makuna/NeoPixelBus/pull/719 merged into master

**Comment by seyfettinkazanir** [comment]
trying to resolve an issue. The led strip is a bit dim and I suspect the first high byte and low byte assigned the probability that I can see on the analyzer won't happen. Is conversion of 8 bit data to 16 bit done in NPB or WLED libraries?. Can I get your opinion on this? @Makuna

**Reply by Makuna** [reply]
Not a WLED expert, but if I remember correctly, they use FastLED for effects and draw directly into the buffer, bypassing SetPixelColor call.

If I am correct, then more than likely they are not taking Endianness into consideration.  A 16bit Value stored in memory on the uC may have the bytes in an order that has the lower 8 bits of the 16bit value actually stored in the upper byte.   This is the case with ESP32.  This means you can't just copy a memory buffer of uint16_t to the buffer exposed by NPB.Pixels() call.  The bytes in the word would be swapped.  

If you look in my library file Neo3WordFeature.h at the applyPixelColor() method, you can see where I solve this issue.

**Reply by blazoncek** [reply]
I am sorry to correct you but WLED uses FastLED only for palettes and some pixel math in some effects. All other stuff is done using NeoPixelBus library (SetPixelColor and (if necessary) GetPixelColor).
We use NeoPixelBusLg class as we depend on global brightness.

Color conversion (in case of USC890x chipset) is done using RgbXXXXColor constructors.

**Reply by Makuna** [reply]
Thanks blazoncek for the clarification.

I think I spot the issue you found.  
Rgb48Color::Dim takes a 16 bit value, but NeoPixelBusLg is only supplying a 8 bit value, so it is always dimmed dramatically.
The question is what is the right solution.  

The "correct" solution is that NeoPixelBusLg be enhanced to support a matching type size to the base color object element size for the luminance.  This would mean that WLED would need to handle having a luminance that goes up to 65535.

A "workable" solution is that NeoPixelBusLg retains the 8bit luminance and scales it based on the base color object element size.

thoughts @blazoncek ?

**Reply by blazoncek** [reply]
If you "demand" 16bit brightness in every situation then it is a simple change to set `bus->SetLuminance(bri<<8);` in case of WLED. I am not sure about other users.

So IMO it would be better to have implicit expansion within NPB in case the target has 16 bit luminance/brightness. I do not know if that is complicated on your side. So I leave final decision to you. ;)

**Reply by Makuna** [reply]
https://github.com/Makuna/NeoPixelBus/pull/720

no changes needed to WLED.
